### PR TITLE
Custom sql generator

### PR DESCRIPTION
**Pull request by Felixim0**

### CODE REVIEW

File: app/customsqlhandler.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+    database_names, fields = await get_database_fields(request)
+    page_number = get_page(request)
+
+    checked_boxes = session.get('custom_sql_previous_checked_fields', [])
+
+    client_input = {}
+    all_input = {}
+    field_classes = []
+    error = ""
```
Comment (SammyPoot) [inline app/customsqlhandler.py]: Single quotes! Also, not sure this needs to be pre-initialized here.

File: app/customsqlchoicehandler.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  async def post(self, request):
+    session = await get_session(request)
+    data = await request.post()
+
+    user_role = await saml.get_role_id(request)
+
+    await saml.ensure_logged_in(request)
+
+    if microservices_permissions(user_role, 'customsql') == False:
+      request['client_ip'] = request.get('client_ip', "No IP Provided")
```
Comment (SammyPoot) [inline app/customsqlchoicehandler.py]: Single quotes!

File: app/customsqlchoicehandler.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    session = await get_session(request)
+    user_role = await saml.get_role_id(request)
+
+    session.pop('custom_sql_previous_filters', None)
+    session.pop('custom_sql_previous_checked_fields', None)
+
+    await saml.ensure_logged_in(request)
+
+    if microservices_permissions(user_role, 'customsql') == False:
+      request['client_ip'] = request.get('client_ip', "No IP Provided")
```
Comment (SammyPoot) [inline app/customsqlchoicehandler.py]: Single quotes!

File: app/customsqlhandler.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    clear = request.match_info['clear']
+    if clear == 'clear':
+      temp = session.get('custom_sql_previous_filters', {})
+      final = {}
+      for key in temp:
+        final[key] = ''
+      session['custom_sql_previous_filters'] = final
+
+    if microservices_permissions(user_role, 'customsql') == False:
+      request['client_ip'] = request.get('client_ip', "No IP Provided")
```
Comment (SammyPoot) [inline app/customsqlhandler.py]: Single quotes!

File: app/customsqlhandler.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+    database_names, fields = await get_database_fields(request)
+    page_number = get_page(request)
+
+    checked_boxes = session.get('custom_sql_previous_checked_fields', [])
+
+    client_input = {}
+    all_input = {}
+    field_classes = []
+    error = ""
```
Comment (SammyPoot) [inline app/customsqlhandler.py]: Single quotes! Also not sure this needs to be initialized here.

File: app/microservice_tables.py [hunk 6]
```diff
@@ -18,6 +18,7 @@ def __init__(self,
                format_as_boolean=False,
                checkbox_value=False,
                name=False,
+               database_association_name="",
```
Comment (SammyPoot) [inline app/microservice_tables.py]: Single quotes!

File: app/microservice_tables.py [hunk 7]
```diff
@@ -363,7 +377,12 @@ def get_table_records(field_classes, json_records, remove_html=False):
             each_record['unique_employee_id'] + '">' + each_record['first_name'] + \
             " " + each_record['surname'] + '</a>'
```
Comment (SammyPoot) [inline app/microservice_tables.py:376]: Single quotes! Also, avoid using backslash line-extension if possible.

File: app/microservice_views.py [hunk 8]
```diff
@@ -113,6 +113,11 @@ def get_views(user_role, microservice_name):
           user_role,
           display_name="Missing Devices",
```
Comment (SammyPoot) [inline app/microservice_views.py]: Single quotes!

File: app/microservice_views.py [hunk 9]
```diff
@@ -113,6 +113,11 @@ def get_views(user_role, microservice_name):
...
       ), )
+  views.append(
+      View("customsql",
```
Comment (SammyPoot) [inline app/microservice_views.py]: Single quotes!

Comment (SammyPoot) [reply app/microservice_views.py]: Also in the three lines below.

File: app/role_matchers.py [hunk 10]
```diff
@@ -73,6 +73,7 @@ def microservices_permissions(role_id, microservice_name):
       "requestlogtable",
       "chromebooktable",
       "missingdevicestable",
+      "customsql",
```
Comment (SammyPoot) [inline app/role_matchers.py]: Single quotes!

File: app/searchfunctions.py [hunk 11]
```diff
@@ -11,6 +11,7 @@
 logger = get_logger('fsdr-ui')
 
 
+#TODO this can probably be removed. Probably. Maybe.
 def get_employee_records_no_device(user_filter=""):
```
Comment (SammyPoot) [inline app/searchfunctions.py]: Single quotes!

File: app/searchfunctions.py [hunk 12]
```diff
@@ -20,6 +21,21 @@ def get_employee_records_no_device(user_filter=""):
...
+  url = URL(FSDR_URL +
+            f'/fieldforce/byMicroservice/customsql/').with_query(user_filter)
+
+  return requests.post(url,
+                       verify=False,
+                       json=all_input,
+                       auth=HTTPBasicAuth(FSDR_USER, FSDR_PASS))
+
+
 def get_microservice_records(endpoint_name, user_filter=""):
```
Comment (SammyPoot) [inline app/searchfunctions.py]: Single quotes!

File: app/microservice_tables.py [hunk 13]
```diff
@@ -351,7 +359,13 @@ def get_fields_to_load(field_classes):
   return (fields_to_load)
 
 
-def get_table_records(field_classes, json_records, remove_html=False):
+def get_table_records(field_classes,
+                      json_records,
+                      remove_html=False,
+                      custom_sql=False):
+  #NEXT  STEPS  -- Add here if the classes are from  CUSTOMSQL make format fsdr_employee_first_name
```
Comment (SammyPoot) [inline app/microservice_tables.py]: Are these comments still needed? Have these steps been finished?

File: app/microservice_views.py [hunk 14]
```diff
@@ -37,7 +37,7 @@ def create_url(self, url, database_name, clear):
     if url == None:
       url = (f'/microservices/{database_name}{clear}')
     else:
-      url = (f'/microservices/{url}{clear}')
+      url = (f'{url}')
```
Comment (SammyPoot) [inline app/microservice_views.py]: Does this do anything?

In python, ('a') normally produces a single value 'a', not a tuple. If you need a tuple, you need a trailing comma: ('a',)

If you only want a string, the else branch would appear to be doing nothing.

**Review by SammyPoot — CHANGES_REQUESTED** [review]

File: app/customsqlhandler.py (hunk 1 above)
Comment (Felixim0) [reply app/customsqlhandler.py]: removed, was  olddebugging   code

File: app/customsqlhandler.py (hunk 5 above)
Comment (Felixim0) [reply app/customsqlhandler.py]: Remved

File: app/microservice_views.py (hunk 14 above)
Comment (Felixim0) [reply app/microservice_views.py]: Ahh, was expecting a string, the brackets must have been copied over from elsewhere erroneously. 
The else branch should set the URL  to whatever the value has been set to (if it was to be overridden  from the default  format above)

File: app/microservice_views.py [hunk 15]
```diff
@@ -35,14 +35,14 @@ def create_filter_label(self, filter_name):
 
   def create_url(self, url, database_name, clear):
     if url == None:
-      url = (f'/microservices/{database_name}{clear}')
+      url = f'/microservices/{database_name}{clear}'
     else:
-      url = (f'{url}')
+      url = f'{url}'
```
Comment (SammyPoot) [inline app/microservice_views.py]: If the brackets aren't needed, this is a no-op. You can remove this entire branch, unless you need to ensure url is a string? If that's the case, pass it to `str` instead.

**Review by SammyPoot — CHANGES_REQUESTED** [review]

File: app/microservice_views.py (hunk 15 above)
Comment (Felixim0) [reply app/microservice_views.py]: Have adjusted

**Review by SammyPoot — APPROVED** [review]

**Pull request merged**